src/KubernetesClient/Watch.cs

```diff
@@ -0,0 +1,482 @@
+        public Task Run(CancellationToken shutdownToken = default)
+        {
+            if (task != null) throw new InvalidOperationException();
+            shutdownToken.ThrowIfCancellationRequested();
+            cts = shutdownToken.CanBeCanceled ?
+                CancellationTokenSource.CreateLinkedTokenSource(shutdownToken) : new CancellationTokenSource();
+            task = Task.Run(Run);
+            return task;
+        }
+
+        async Task Run()
+        {
+            while(!cts.IsCancellationRequested)
+            {
+                bool wasOpen = false; // used to determine whether we should raise OnClosed
+                try
+                {
+                    // when watching multiple items when we don't know the starting version, we first have to retrieve the list and then
+                    // watch for changes. if we just started watching, we'd get an initial mishmash of resource versions and wouldn't be
+                    // able to know what the latest version is. when listing, the appropriate version is part of the list metadata
+                    KubernetesList<T> list = null;
+                    if (IsListWatch && LastVersion == null)
+                    {
+                        using (KubernetesResponse listResponse = await req.WatchVersion(null).ExecuteAsync(cts.Token).ConfigureAwait(false))
+                        {
+                            if (listResponse.IsError)
+                            {
+                                await ReportErrorAndWait(listResponse).ConfigureAwait(false);
+                                continue;
+                            }
+                            list = await listResponse.GetBodyAsync<KubernetesList<T>>(failIfEmpty: true).ConfigureAwait(false);
+                            LastVersion = list.Metadata?.ResourceVersion;
+                        }
+                    }
+
+                    // now start the watch
+                    using (KubernetesResponse response = await req.WatchVersion(LastVersion ?? "").ExecuteAsync(cts.Token).ConfigureAwait(false))
+                    {
+                        if (response.IsError) // if the watch request failed...
+                        {
+                            // if our resource version is too old, reset it. otherwise report the error
+                            if (response.StatusCode == HttpStatusCode.Gone) LastVersion = null; // NOTE: in practice this seems to be reported as a watch event...
+                            else await ReportErrorAndWait(response).ConfigureAwait(false);
+                        }
+                        else // if the response was successful, then the watch is open...
+                        {
+                            bool reset = LastVersion == null || list != null;
+                            if (reset && Reset != null) Reset(this); // if we may have lost events, let the client know
```

Comment:
There is a race condition here between the null check, `remove_Reset`, and the event being invoked.

Consider invoking all events using the null-conditional operator to eliminate the race condition.

Reply:
Well, I really don't think people should be adding and removing event handlers on a running watch, since that's bound to create race conditions in whether or not they receive the events. But sure, just as a best practice to guard against that pathological case, I'll make the change.